### PR TITLE
Improving lstm code, solving issue with activation table generation, …

### PR DESCRIPTION
**Pull request by etfortin**
Improving lstm code , solving issue with activation table generation,solving issues with tb

### CODE REVIEW

File: hls4ml/model/hls_layers.py [hunk 1]
```diff
@@ -899,15 +899,14 @@ def initialize(self):
         data  = self.model.get_weights_data(self.name, 'kernel')
         data2 = self.model.get_weights_data(self.name, 'recurrent_kernel')
         data3 = self.model.get_weights_data(self.name, 'bias')
-        print("data3:")
-        print(data3)
         #print(self.get_layers.get_weights())
         weight_types=["i","f","c","o"]
         for i in range (0,4):
           self.add_weights_variable(name='weight_%s'% weight_types [i], var_name='kernel_%s_{index}' % weight_types [i], data=data[0][i*self.get_attr('n_in'):(i+1)*(self.get_attr('n_in'))], quantizer=self.get_attr('weight_quantizer'), compression=None)
```
Comment (laurilaatu) [inline hls4ml/model/hls_layers.py]: Should use `'.. {} ..'.format(..)`

File: hls4ml/templates/quartus/firmware/nnet_utils/lstm_cell.h [hunk 2]
```diff
@@ -157,75 +157,106 @@ void lstm_cell(
...
-         data_T c[CONFIG_T::n_in] hls_register;
-         data_T o[CONFIG_T::n_in] hls_register;
-         data_T h[CONFIG_T::n_in] hls_register;
-    
+        data_T gate_i[lstm_config::n_in] hls_register;
+        data_T gate_f[lstm_config::n_in] hls_register;
+        data_T gate_c[lstm_config::n_in] hls_register;
+        data_T gate_o[lstm_config::n_in] hls_register;
+        data_T gate_ic[lstm_config::n_in] hls_register;
+        data_T gate_forgot[lstm_config::n_in] hls_register;
```
Comment (laurilaatu) [inline hls4ml/templates/quartus/firmware/nnet_utils/lstm_cell.h]: The gate is called forget

File: hls4ml/templates/quartus/firmware/nnet_utils/nnet_activation.h [hunk 3]
```diff
@@ -99,12 +99,13 @@ void  relu1(data_T data[CONFIG_T::n_in], res_T res[CONFIG_T::n_in])
...
 void  sigmoid(data_T data[CONFIG_T::n_in], res_T res[CONFIG_T::n_in])
 {
+    int MAX_TABLE_value=8;
     #include "activation_tables/sigmoid_table.tb"
     // Index into the lookup table based on data
     #pragma unroll
     for (int ii=0; ii<CONFIG_T::n_in; ii++) {
         int data_round = (data[ii]*(CONFIG_T::table_size/16)).to_int();
-        int index = data_round + 8*CONFIG_T::table_size/16;
+        int index = data_round + MAX_TABLE_value *CONFIG_T::table_size/16;
```
Comment (laurilaatu) [inline hls4ml/templates/quartus/firmware/nnet_utils/nnet_activation.h:108]: Should also include the minimum value and `table_size/16` should become `table_size/(MAX_TABLE_value-MIN_TABLE_value)` same applies in other places too where `table_size/16` is used.

File: hls4ml/writer/quartus_writer.py [hunk 4]
```diff
@@ -575,20 +580,23 @@ def write_activation_tables(self, model):
             in_val = 2*8.0*(i-float(table_size)/2.0)/float(table_size)
```
Comment (laurilaatu) [inline hls4ml/writer/quartus_writer.py:582]: Table range is now hardcoded to be from -8 to 8, and for `tanh` the range is from -4 to 4 when writing the table

File: hls4ml/templates/quartus/firmware/nnet_utils/nnet_activation.h (hunk 3 above)
Comment (etfortin) [reply hls4ml/templates/quartus/firmware/nnet_utils/nnet_activation.h:108]: I will soon rewrite all code for activation function with the optimisation I made so for now,I will let like this

File: hls4ml/writer/quartus_writer.py (hunk 4 above)
Comment (etfortin) [reply hls4ml/writer/quartus_writer.py:582]: This code was not modified by me and is working , we will not fully rewrite hls4ml

**Pull request merged**